### PR TITLE
go_1_18: Use apple_sdk_11_0.callPackage

### PR DESCRIPTION
**Pull request by toonn**
Go 1.18 requires a newer SDK than the one we build from sources. As a
workaround we're making use of the SDK we're using for aarch64-darwin.
This means Go 1.18 will not work on any Darwin systems that don't have
forwards-compatible SDK versions with the particular package in
question. We might need to mark Go packages broken based on the macOS
version rather than just the platform and architecture.

Until we find a better solution, Go packages will need to make sure to
get all their (Darwin) system dependencies from the `apple_sdk_11_0`,
this includes dependencies of build tools like `xcbuild`.

For convenience `darwin.apple_sdk_11_0` has a `callPackage` attribute
which provides the correct `stdenv` and `xcbuild` attributes as
arguments. This function can be expanded to substitute other necessary
arguments when they come up.

###### Description of changes

Use `darwin.apple_sdk_11_0.callPackage` for `go_1_18`, `buildGo118Module` and `buildGo118Package` to insert a stdenv and `xcbuild` based on `apple_sdk_11_0` on Darwin.

I've merged #179622 into this so I can run a [Hydra evaluation for x86_64-darwin](https://hydra.nixos.org/eval/1770591). I've kept this as a draft so I can either rebase after #179622 is merged or change the target branch to staging because of all the rebuilds.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] **Same change but not this exact patch, [ran by @reckenrode](https://github.com/NixOS/nixpkgs/pull/176661#issuecomment-1177872920)** Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -13407,8 +13360,8 @@ with pkgs;
     buildPackages = buildPackages // { stdenv = buildPackages.gcc8Stdenv; };
   });
 
-  go_1_18 = callPackage ../development/compilers/go/1.18.nix {
-    inherit (darwin.apple_sdk.frameworks) Security Foundation;
+  go_1_18 = darwin.apple_sdk_11_0.callPackage ../development/compilers/go/1.18.nix {
+    inherit (darwin.apple_sdk_11_0.frameworks) Security Foundation;
   };
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:13365]: I wouldn't mind a comment here

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -21753,20 +21700,20 @@ with pkgs;
   buildGo117Package = callPackage ../development/go-packages/generic {
     go = buildPackages.go_1_17;
   };
-  buildGo118Package = callPackage ../development/go-packages/generic {
+  buildGo118Package = darwin.apple_sdk_11_0.callPackage ../development/go-packages/generic {
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:21703]: or here

File: pkgs/top-level/all-packages.nix [hunk 3]
```diff
@@ -21753,20 +21700,20 @@ with pkgs;
...
   };
 
-  buildGoPackage = buildGo117Package;
+  buildGoPackage = buildGo118Package;
 
   buildGo117Module = callPackage ../development/go-modules/generic {
     go = buildPackages.go_1_17;
   };
-  buildGo118Module = callPackage ../development/go-modules/generic {
+  buildGo118Module = darwin.apple_sdk_11_0.callPackage ../development/go-modules/generic {
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:21712]: or here

**Review by SuperSandro2000 — COMMENTED** [review]
I think the easiest is that I cherry-pick the darwin fix commit into my branch when it is done because I need to rebase that on master again to remove new occurrences of buildGo118Module

**Comment by reckenrode** [comment]
The `callPackages` changes LGTM, though I agree regarding adding a comment to explain why the Darwin `callPackage` is being used. Should the Go-using packages that are marked broken on x86_64-darwin also have that removed in this PR?

> We might need to mark Go packages broken based on the macOS version rather than just the platform and architecture.

Is this possible? I can see needing something like that in the future for DXVK when MoltenVK makes greater use of Metal 3 (and presumably can provide the missing extensions on macOS 13 or newer).

**Comment by SuperSandro2000** [comment]
> Should the Go-using packages that are marked broken on x86_64-darwin also have that removed in this PR?

Please lets do this in another PR. I don't want the go 1.18 PR to completely grow out of hand. There might also be failing tests or go.sum/vendoring files which need updating.

**Comment by SuperSandro2000** [comment]
Cherry-picked into #179622

**Pull request closed without merging**